### ISSUE TITLE
Last PR seems to have broken `text/csv` MIME type.

### ISSUE THREAD

**Issue by gwright99** [seqeralabs/datasets-automation-blog#4]
Users trying to implement the Lambda after the merging of [Fix python install #2](https://github.com/seqeralabs/datasets-automation-blog/pull/2/files) report Tower throwing a MIME type error when the Lambda tries to create a Tower dataset: `ERROR: Given file is not a dataset file. Detected media type: 'application/octet-stream. Allowed types: 'text/csv, text/tab-separated-values'.`

## Investigation
- The reporting site uploaded the CSV file from a Windows-based machine. This machine had proper MIME types set in the Registry.
- S3 recognized the file as a CSV type after upload.
- The same file could be uploaded directly to Tower successfully.
- The error was only thrown when executing the Lambda.

Image change:
- The merge PR changes how the docker image is built: rather than using a base OS image and installing python via package manager, it now uses an image with Python3 pre-baked.
- It appears that `python:3*slim*` images [do not ship with some MIMEtype packages](https://github.com/docker-library/python/issues/254). 

Image peculiarities:
- Thinking the `slim` tag was the problem, I originally tried with `FROM python:3.9`.  Unfortunately, this resulted in a bizarre error from the Lambda RIE: 
```log
<html>
<body>
<!--StartFragment-->
2024-10-31T19:21:11.927Z | INIT_REPORT Init Duration: 504.06 ms Phase: init Status: error Error Type: Runtime.ExitError
-- | --
  | 2024-10-31T19:23:28.919Z | Traceback (most recent call last):
  | 2024-10-31T19:23:28.919Z | File "<frozen runpy>", line 198, in _run_module_as_main
  | 2024-10-31T19:23:28.919Z | File "<frozen runpy>", line 88, in _run_code
  | 2024-10-31T19:23:28.919Z | File "/var/task/awslambdaric/__main__.py", line 20, in <module>
  | 2024-10-31T19:23:28.920Z | main(sys.argv)
  | 2024-10-31T19:23:28.920Z | File "/var/task/awslambdaric/__main__.py", line 16, in main
  | 2024-10-31T19:23:28.920Z | bootstrap.run(app_root, handler, lambda_runtime_api_addr)
  | 2024-10-31T19:23:28.920Z | File "/var/task/awslambdaric/bootstrap.py", line 399, in run
  | 2024-10-31T19:23:28.920Z | event_request = lambda_runtime_client.wait_next_invocation()
  | 2024-10-31T19:23:28.920Z | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  | 2024-10-31T19:23:28.920Z | File "/var/task/awslambdaric/lambda_runtime_client.py", line 68, in wait_next_invocation
  | 2024-10-31T19:23:28.920Z | response_body, headers = runtime_client.next()
  | 2024-10-31T19:23:28.920Z | ^^^^^^^^^^^^^^^^^^^
  | 2024-10-31T19:23:28.920Z | AttributeError: 'NoneType' object has no attribute 'next'

<!--EndFragment-->
</body>
</html>
```

- The execution script runs the following line when local: `exec /usr/local/bin/aws-lambda-rie-x86_64 /usr/bin/python3 -m awslambdaric "$@"`
- The 3.9 image has 2 different python versions in -- is it possible installing a `python3-pip` package is causing the problem? **Spoiler: YES**.

```  bash
# In image based on: FROM python3.9

root@16e63a02836f:/var/task# python3 --version
Python 3.9.17

root@16e63a02836f:/var/task# /usr/bin/python3
Python 3.11.2 (main, Aug 26 2024, 07:20:54) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>>
```

Only one of the python instances has `runtime_client` available:
```bash
# In image based on: FROM python3.9

root@d5f823346ad5:/var/task# python3
Python 3.9.17 (main, Jun 14 2023, 18:56:13)
[GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import runtime_client
>>>
>>> quit()

root@d5f823346ad5:/var/task# /usr/bin/python3
Python 3.11.2 (main, Aug 26 2024, 07:20:54) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import runtime_client
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'runtime_client'
```

However, when the same transaction is run from a base image (e.g. Ubuntu) which Python is then installed onto:
```bash
# In image based on: FROM ubuntu:22.04

root@f5c2e73b37e3:/var/task# python3
Python 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> quit()
root@f5c2e73b37e3:/var/task# /usr/bin/python3
Python 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import runtime_client
>>>
```

If I patch `entry_script.sh` to this, the Python3.9 image starts working (locally; I assume same has to be done for AWS):
```bash
#!/bin/sh

# Reference: https://docs.aws.amazon.com/lambda/latest/dg/images-test.html
if [ -z "${AWS_LAMBDA_RUNTIME_API}" ]; then
  # OLD: exec /usr/local/bin/aws-lambda-rie-x86_64 /usr/bin/python3 -m awslambdaric "$@"
  exec /usr/local/bin/aws-lambda-rie-x86_64 python3 -m awslambdaric "$@"
else
  exec /usr/bin/python3 -m awslambdaric "$@"
fi
```

## Smoking Gun
Thanks to eagle-eyed @endre-seqera , we noticed that the PR also removed `update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.9 1` from the original Dockerfile. This would explain the 3.9 / 3.11 discrepancy I encountered during the most recent testing re: the Python3.9 image. 

With this in mind, we can see that the Python3.9 image actually has the two different versions in different places in the image:
```bash
# In image based on: FROM python3.9

root@860df6123609:/var/task# ls -al /usr/bin/ | grep -i python
lrwxrwxrwx 1 root root         24 Aug 26 07:20 pdb3.11 -> ../lib/python3.11/pdb.py
lrwxrwxrwx 1 root root         31 Apr  9  2023 py3versions -> ../share/python3/py3versions.py
lrwxrwxrwx 1 root root         10 Apr  9  2023 python3 -> python3.11
-rwxr-xr-x 1 root root    6839896 Aug 26 07:20 python3.11

root@860df6123609:/var/task# which python3
/usr/local/bin/python3                                                   # <-- Note `local`

root@860df6123609:/var/task# ls -al /usr/local/bin | grep -i python
lrwxrwxrwx 1 root root        7 Jun 14  2023 python -> python3
lrwxrwxrwx 1 root root       14 Jun 14  2023 python-config -> python3-config
lrwxrwxrwx 1 root root        9 Jun 14  2023 python3 -> python3.9
lrwxrwxrwx 1 root root       16 Jun 14  2023 python3-config -> python3.9-config
-rwxr-xr-x 1 root root    17592 Jun 14  2023 python3.9
-rwxr-xr-x 1 root root     3045 Jun 14  2023 python3.9-config
```

## Explanation of Problem Occurence
Doing a fresh install of Python on `debian:stable-slim` (_the original image used when the blog was first published_), results in the binary being placed in `/usr/bin/python3.xx`. I wrote the rest of the configuration / content expecting this to be true. 

The image swap that occurred in the PR used an image where `python3` != `/usr/bin/python`, and the removal of the `install-alternatives` failed to normalize the command. This in turn would affect package installation since we are using `pip3` (which depends on Python), broke submission capabilities.

This behaviour is confirmed with two tests:

1. Dumping environment variables in AWS confirms that the `AWS_LAMBDA_RUNTIME_API` was found.
![image](https://github.com/user-attachments/assets/71dc7b92-8e1c-48e2-8f24-b3b95e96d664)

2. Exec-ing into a local copy of the container, adding `pandas` to _requirements.txt_, and then running `pip3 install -r requirements.txt`.
```bash
# In image based on: FROM python3.9

root@030c6bc15728:/var/task# python3
Python 3.9.17 (main, Jun 14 2023, 18:56:13)
[GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pandas
>>> quit()

root@030c6bc15728:/var/task# /usr/bin/python3
Python 3.11.2 (main, Aug 26 2024, 07:20:54) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pandas
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'pandas
```

## Remediation Approach
There are a few ways to remediate this (_all working for @gwright99, still waiting on confirmation by client_):

1. Revert to the old technique (generic base OS + `apt install` + `update-alternatives`). Ensure that `/usr/bin/python3` points to the same binary as `python3`

2. Remove `/usr/bin/` from the `/usr/bin/python3` entries in the _entry_script.sh_. This will align binary execution with package install. 

Both options are feasible:
- I tend to favour (1) since it's more explicit and aligns to the original content. The downside is that the image requires more set-up; but this is also not necessarily bad since we know exactly what's going into the image.

- Conversely, using the generic `python3` invocation should make it easier to update source images. This would avoid the problem since when setting up **ubuntu:22.04** with **Python3.9** -- the version was so old, `apt` could no longer find it and additional commands had to be added to include the **deadsnakes ppa** in the Docker image.

**Comment by gwright99** [comment]
@schaluva -- any thoughts on this issue before I make a decision and get this problem resolved?

**Comment by gwright99** [comment]
**Update:** This actually ended up being two separate problems:

1. A problem with how `python3` was being invoked in the created Docker container.
2. A problem re: TW CLI MIME exception when trying to create a dataset via the Lambda.

The MIME portion has been resolved via #5 .
This issue will handle the Docker image fix.

**Comment by gwright99** [comment]
As per the associated PR, I decided to implement the fix as follows:

1. Use a `FROM python:3.9` base. 
    This means there's a more consistent approach for implementers while avoiding different OS flavour differences (e.g. Debian not having python3.9 available from `apt`, Ubuntu requiring the addition of the deadsnakes ppa, etc.)

2. To accommodate the potential there are multiple versions of Python in the running image, use the relative `python3` command, rather than hardcoded path to `usr/bin/python3`.  This also allowed me to remove the `update-alternatives` command in the Dockerfile.

These changes should make the solution flexibile enough to accommodate modern use despite the 2+ years since the content was originally written.
